EIPS/eip-7851.md

```diff
@@ -0,0 +1,223 @@
+discussions-to: https://ethereum-magicians.org/t/eip-7851-private-key-deactivation-and-reactivation-of-eoas-with-delegated-code/22344
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-12-27
+requires: 20, 2612, 7701, 7702
+---
+
+## Abstract
+
+This EIP introduces a precompiled contract that enables Externally Owned Accounts (EOAs) with delegated control to smart contracts via [EIP-7702](./eip-7702) to deactivate/reactivate their private keys. This design does not require additional storage fields or account state changes. By leveraging delegated code, reactivation can be performed securely through mechanisms such as social recovery.
+
+## Motivation
+
+[EIP-7702](./eip-7702) enables EOAs to gain smart contract capabilities, but the private key of the EOA still retains full control over the account.
+
+With this EIP, EOAs can fully migrate to smart contract wallets, while retaining recovery options with reactivation. The flexible deactivate/reactivate design also paves the way for native account abstraction. e.g., [EIP-7701](./eip-7701).
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Parameters
+
+| Constant                          | Value                |
+|-----------------------------------|----------------------|
+| `PRECOMPILE_ADDRESS`              | `0xTBD`              |
+| `PRECOMPILE_GAS_COST`             | `5000` (tentative)   |
+
+### Delegated code encoding
+
+The deactivation status is encoded by appending or removing the `0x00` byte at the end of the delegated code. The transitions between two states are as follows:
+
+- Active state: `0xef0100 || address`, the private key is active and can sign transactions.
+- Deactivated state: `0xef0100 || address || 0x00`, the private key is deactivated and cannot sign transactions.
+
+### Precompiled contract
+
+A new precompiled contract is introduced at address `PRECOMPILE_ADDRESS`. For each call, it consumes `PRECOMPILE_GAS_COST` gas, and the precompiled contract executes the following steps:
+
+- The precompiled contract checks that the caller is an EOA with delegated code (i.e., its account code begins with the prefix `0xef0100`, as defined in [EIP-7702](./eip-7702)). If the code does not conform to the required prefix, the contract MUST terminate without making any state changes.
+- The precompile determines the current state of the delegated code based on its byte length:
+  - If the delegated code is 24 bytes (`0xef0100 || address || 0x00`), it removes the last byte (`0x00`), transitioning to the active state (`0xef0100 || address`).
+  - If the delegated code is 23 bytes (`0xef0100 || address`), it appends `0x00`, transitioning to the deactivated state (`0xef0100 || address || 0x00`).
+- The updated delegated code is saved as the new account code for the EOA.
```

Comment:
I think a rule should be added here. In case that we are updating the delegated code, but we are running in "static" mode (so a call chain after using `STATICCALL`) this should abort (revert, no state-changes).

Reply:
yes. you're right! I added the `STATICCALL` error case in this commit: [0f2aaba](https://github.com/ethereum/EIPs/pull/9193/commits/0f2aabaad752324ad8bab51a123213ca290a1c10)